### ISSUE TITLE
Unable to get PID from real parent

### ISSUE THREAD

**Issue by calavera** [iovisor/bcc#1883]
I was working on porting execnoop to Go and I wanted to get the parent PID from the current task. Unfortunately, the TGID for the task's real parent is always 0.

I used the code from this closed issue: https://github.com/iovisor/bcc/issues/799

```
  struct task_struct *task;
  struct task_struct *real_parent_task;
  u64 ppid;

  task = (struct task_struct *)bpf_get_current_task();
  bpf_probe_read(&real_parent_task, sizeof(real_parent_task), &task->real_parent);
  bpf_probe_read(&ppid, sizeof(ppid), &real_parent_task->tgid);
```

I also used `bpf_trace_printk` to print the result from calling `bpf_probe_read` and it always returns 0. You can see my bpf program in this PR: https://github.com/iovisor/gobpf/pull/133/files#diff-3ef2d414e040c89dc6f8a62e402b33a9R92

**Comment by pchaigno** [comment]
You wouldn't happen to run a 4.13 Linux kernel by any chance?

**Comment by calavera** [comment]
Yes! Is that a known issue? I couldn't find anything related anywhere:

```
Linux c 4.13.0-46-generic #51-Ubuntu SMP Tue Jun 12 12:36:29 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by palmtenor** [comment]
@calavera Can you try the fix mentioned here: https://github.com/iovisor/bcc/issues/1548#issuecomment-359964600?

**Comment by calavera** [comment]
@palmtenor that doesn't seem to work in my case. I applied this patch to execsnoop:

```diff
diff --git i/examples/bcc/execsnoop/execsnoop.go w/examples/bcc/execsnoop/execsnoop.go
index 1bb3f20..50687c9 100644
--- i/examples/bcc/execsnoop/execsnoop.go
+++ w/examples/bcc/execsnoop/execsnoop.go
@@ -37,6 +37,9 @@ const (
 )
 
 const source string = `
+#define randomized_struct_fields_start  struct {
+#define randomized_struct_fields_end    };
+
 #include <uapi/linux/ptrace.h>
 #include <linux/sched.h>
 #include <linux/fs.h>
@@ -272,9 +275,9 @@ func main() {
                                        RetVal: event.RetVal,
                                }
 
-                               if event.Ppid == 0 {
-                                       event.Ppid = getPpid(event.Pid)
-                               }
+                               //                              if event.Ppid == 0 {
+                               //                                      event.Ppid = getPpid(event.Pid)
+                               //                              }
 
                                if event.Ppid != 0 {
                                        p.Ppid = strconv.FormatUint(event.Ppid, 10)
```

And I don't get any PID back:

![screenshot from 2018-07-10 14-55-22](https://user-images.githubusercontent.com/1050/42539883-4fddd8ee-8451-11e8-910c-49607c572422.png)

**Comment by calavera** [comment]
Btw, worth noticing that my implementation of execsnoop reads from /proc as a fallback. So if this problem is not an issue in other kernels, it will read the PID from the real parent correctly, but i'll test it in different kernels.

**Comment by pchaigno** [comment]
Could you try to reproduce the issue with bcc's version of execsnoop?

**Comment by calavera** [comment]
@pchaigno I made the same change in bcc's version in #1885. I see the same behavior with 4.13. I have not tested in newer kernels yet.

**Comment by calavera** [comment]
I'm going to close this issue as solved now that #1885 is in master. Thanks a lot for your feedback! :raised_hands:
